File: src/main/java/com/shelterhub/dto/request/AnimalRequest.java
package com.shelterhub.dto.request;

import com.shelterhub.domain.enums.AnimalType;
import com.shelterhub.domain.enums.Gender;
import com.shelterhub.domain.enums.Size;
import com.shelterhub.domain.model.Animal;
import lombok.Builder;
import lombok.Data;
import org.apache.commons.lang3.EnumUtils;

import java.util.UUID;

@Data
@Builder
public class AnimalRequest {
    private UUID id;
    private String name;
    private String identification;
    private String behavior;
    private String breed;
    private String history;
    private EstimatedAgeRequest estimatedAge;
    private UUID medicalRecordId;
    private String animalType;
    private Gender gender;
    private Size size;

    public Animal toAnimal() {
        return Animal.builder()
                .id(getId())
                .name(getName())
                .identification(getIdentification())
                .estimatedAge(estimatedAge.toEstimatedAge())
                .medicalRecordId(getMedicalRecordId())
                .animalType(EnumUtils.getEnum(AnimalType.class, getAnimalType(), AnimalType.Unknown))
                .behavior(getBehavior())
                .breed(getBreed())
                .history(getHistory())
                .gender(getGender())
                .size(getSize())
                .build();
    }
}

File: src/main/kotlin/br/com/shelterhubmanagementapi/dto/request/AnimalRequest.kt
package br.com.shelterhubmanagementapi.dto.request

import br.com.shelterhubmanagementapi.domain.enums.AnimalType
import br.com.shelterhubmanagementapi.domain.enums.Gender
import br.com.shelterhubmanagementapi.domain.enums.Size
import br.com.shelterhubmanagementapi.domain.model.Animal
import java.lang.NullPointerException
import java.util.UUID

data class AnimalRequest(
    val id: UUID = UUID.randomUUID(),
    val name: String,
    val identification: String,
    val behavior: String,
    val breed: String,
    val history: String,
    val estimatedAge: EstimatedAgeRequest? = null,
    val medicalRecordId: UUID?,
    val animalType: String,
    val gender: Gender,
    val size: Size,
)

fun AnimalRequest.toAnimal() =
    Animal(
        id = this.id,
        name = this.name,
        identification = this.identification,
        behavior = this.behavior,
        breed = this.breed,
        history = this.history,
        estimatedAge = this.estimatedAge?.toLocalDate(),
        medicalRecordId = this.medicalRecordId ?: throw NullPointerException(),
        animalType = AnimalType.getOrDefault(this.animalType),
        gender = this.gender,
        size = this.size,
    )
